packages/editor/src/editor/plugins/createWithInsertData.ts

```diff
@@ -273,11 +274,35 @@ function _regenerateKeys(
   fragment: Descendant[],
   keyGenerator: () => string,
   spanTypeName: string,
+  editorTypes: PortableTextMemberSchemaTypes,
 ): Descendant[] {
   return fragment.map((node) => {
     const newNode: Descendant = {...node}
     // Ensure the copy has new keys
     if (editor.isTextBlock(newNode)) {
+      const annotations = editorTypes.annotations.map((t) => t.name)
+
+      // Ensure that if there are no annotations, we remove the markDefs
+      if (annotations.length === 0) {
+        const {markDefs, ...NewNodeNoDefs} = newNode
+
+        return {...NewNodeNoDefs, _key: keyGenerator()}
+      }
+
+      // Ensure that all annotations are allowed
+      const hasForbiddenAnnotations = (newNode.markDefs || []).some((def) => {
+        return !annotations.includes(def._type)
+      })
+
+      // if they have forbidden annotations, we remove them and keep the rest
+      if (hasForbiddenAnnotations) {
+        const allowedAnnotations = (newNode.markDefs || []).filter((def) => {
+          return annotations.includes(def._type)
+        })
+
+        return {...newNode, markDefs: allowedAnnotations, _key: keyGenerator()}
```

Comment:
I find this function a bit confusing to read since this if clause now either does an early return of manipulates `newNode` directly. But maybe that's not something we should refactor unless the rest of the function's branches are covered by tests as well.

Reply:
I think it does what it is expected,
the new node will always be modified directly (even in the old part of the code that wasn't changed). The point is to make sure that whatever node is returned from this has the correct MarkDefs.

What I am wondering now is if in the case where there are two annotations, one is allowed and one isn't if the allowed one should just be allowed to proceed or should receive the same treatment as the line bellow since as the code is now that will not happen now that I think about it 🤔 @skogsmaskin thoughts?

Reply:
True, I too think it does what is expected. I just found it a tad hard to read and understand the logic 😇

Reply:
> What I am wondering now is if in the case where there are two annotations, one is allowed and one isn't if the allowed one should just be allowed to proceed

I think all allowed annotations should be preserved. Good point! We probably need a test case for that and adjust the logic. If you are busy then I can give it a try for you if you like :)

Reply:
> True, I too think it does what is expected. I just found it a tad hard to read and understand the logic 😇

It's an absolutely fair point, should have definitely added a comment about it to make it clearer :)

Reply:
> > What I am wondering now is if in the case where there are two annotations, one is allowed and one isn't if the allowed one should just be allowed to proceed
> 
> I think all allowed annotations should be preserved. Good point! We probably need a test case for that and adjust the logic. If you are busy then I can give it a try for you if you like :)

I'm currently in the middle of something else, if you have the time I'd be super thankful since i'll likely only have time on Monday 🙏

Reply:
I'll give it a try! I really should get more familiar with this codebase anyway.

Reply:
If you need help let me know I can try my best even if I myself don't have a lot of context :)